### PR TITLE
JSpecify generics checks for conditional expressions

### PR DESCRIPTION
**Pull request by NikitaAware**
For a conditional expression _e_, the type parameter nullability annotations of the true-subpart and false-subpart should match the annotations required by _e_'s context.  E.g., if _e_ is the right-hand side of an assignment, the annotations should match those for the type parameters for the assignment's left-hand side:
```java
// this is fine
A<@Nullable String> t = condition? new A<@Nullable String> : new A<@Nullable String>();
// this is bad
A<@Nullable String> t2 = condition? new A<String> : new A<String>();
// this is also bad
A<@Nullable String>  t1 = condition ? new A<String>(): new A<@Nullable String>()
```
Other contexts include returns, parameter passing, and being nested inside another conditional expression.  In our experience, it seems that `javac` captures the type parameter nullability required by _e_'s context in the type of _e_ itself.  So, our handling of conditional expressions simply checks that the types of both the true and false sub-expressions of _e_ are assignable to (i.e., a subtype of) the type of _e_, using the same machinery for checking assignments from #715.

### CODE REVIEW

**Comment by msridhar** [comment]
@lazaroclapp I've already done some review of this one.  Whenever you get time, it'd be great if you could take a look.

File: nullaway/src/test/java/com/uber/nullaway/NullAwayJSpecifyGenericsTests.java [hunk 1]
```diff
@@ -557,6 +557,107 @@ public void genericFunctionReturnTypeMultipleReturnStatementsIfElseBlock() {
...
+  public void genericsChecksForTernaryOperator() {
+    makeHelper()
+        .addSourceLines(
+            "Test.java",
+            "package com.uber;",
+            "import org.jspecify.annotations.Nullable;",
+            "class Test {",
+            "static class A<T extends @Nullable Object> { }",
+            "  static A<String> testPositive(boolean t) {",
+            "    // BUG: Diagnostic contains: Conditional expression",
```
Comment (lazaroclapp) [inline nullaway/src/test/java/com/uber/nullaway/NullAwayJSpecifyGenericsTests.java:570]: Can we at least use "Conditional expression must have type" as the match string here, just matching on "Conditional expression" is very ambiguous unless someone has in mind the error's full text already.

**Comment by lazaroclapp** [comment]
Just finished a pass above. Overall this looks good to me, with some nits/comments. Sorry once again for the delay, but I was sick with COVID all of last week (still recovering now).

File: nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java [hunk 2]
```diff
@@ -366,4 +384,51 @@ private Type.ClassType typeWithPreservedAnnotations(ParameterizedTypeTree tree)
...
+    Type falsePartType = getTreeType(falsePartTree);
+    // The condExpr type should be the least-upper bound of the true and false part types.  To check
+    // the nullability annotations, we check that the true and false parts are assignable to the
+    // type of the whole expression
+    if (condExprType instanceof Type.ClassType) {
+      if (truePartType instanceof Type.ClassType) {
+        if (!compareNullabilityAnnotations(
+            (Type.ClassType) condExprType, (Type.ClassType) truePartType)) {
+          reportMismatchedTypeForTernaryOperator(
+              truePartTree, condExprType, truePartType, state, analysis);
```
[the remark this answers is not in the document]
Comment (msridhar) [reply nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java:422]: You're right that we could early return and only report one error there.  However, I think what we would really want for that test case is an error message saying the conditional expression as a whole has type `A<String>`, but the context (i.e., the left-hand side of the assignment) requires type `A<@Nullable String>` which is mismatched.  If we early return, we will report the error for the true part of the conditional expression, but not for the expression as a whole, which I think is still not ideal.  If you think just reporting for the true part is better than what we do now, we can add the early return.

File: nullaway/src/test/java/com/uber/nullaway/NullAwayJSpecifyGenericsTests.java [hunk 3]
```diff
@@ -557,6 +557,107 @@ public void genericFunctionReturnTypeMultipleReturnStatementsIfElseBlock() {
...
+            "    A<@Nullable String> t1 = t ? new A<String>() : new A<@Nullable String>();",
+            "    // BUG: Diagnostic contains: Conditional expression",
+            "    return t ? new A<@Nullable String>() : new A<@Nullable String>();",
```
[the remark this answers is not in the document]
Comment (NikitaAware) [reply nullaway/src/test/java/com/uber/nullaway/NullAwayJSpecifyGenericsTests.java:573]: This code should handle all the cases where we have conditional expressions. I will add a test case that passes the ternary expression as an argument. Thanks for pointing this out.

**Comment by msridhar** [comment]
@NikitaAware your recent commits were pushed with a wrong email address, so GitHub got confused about the CLA.  I tried  a force-push to fix this; we'll see if it works.  I recommend deleting your local branch and pulling from the remote before making any further changes.

File: nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (lazaroclapp) [reply nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java:420]: Makes sense to keep this as is for now and working on a better reporting location as a follow up, then, I think...

**Review by lazaroclapp — APPROVED** [review]
LGTM!

**Pull request merged**